Improve documentation on shellhooks (#2027)
theforeman/foreman-documentation: guides/common/modules/con_shellhooks-plugin.adoc

For advanced integrations, where a single shell script can contain multiple commands, you can install a {SmartProxy} shellhooks plugin that exposes executables using a REST HTTP API.

You can then configure a webhook to reach out to a {SmartProxy} API to run a predefined shellhook.
The shellhook can for example contain commands or edit files.

You must place the scripts in `/var/lib/foreman-proxy/shellhooks` as executables with only alphanumeric characters and underscores in the name.

Because the shellhook script only uses command line arguments, the HTTPS payload can be empty (the *Empty Payload* webhook template).
You can provide command line arguments using X-Shellhook-Arg-1 to X-Shellhook-Arg-99.
For more information on passing arguments to shellhook script, see:

* xref:passing-arguments-to-shellhook-script-using-webhooks_{context}[]
* xref:passing-arguments-to-shellhook-script-using-curl_{context}[]

The HTTP method must be POST.
An example URL would be: `\https://{smartproxy-example-com}:{smartproxy_port}/shellhook/My_Script`.

[NOTE]
====
Unlike the `shellhooks` directory, the URL must contain `/shellhook/` in singular to be valid.
====

> You must enable *{SmartProxy} Authorization* for each webhook that is connected to a shellhook to enable it to authorize a call.

Reviewer: ```suggestion
You must enable *{SmartProxy} Authorization* for each webhook connected to a shellhook to enable it to authorize a call.
```

[id="shellhooks_{context}"]
= Shellhooks

> With webhooks, you can only map one {Project} event to one API call.

Reviewer 2: I am confused. The title is "Shellhooks", but you're immeditely starting with "webhooks".

Author: Yes, this is to show the difference between webhooks and shellhooks. Shellhooks improve on webhooks because they can map multiple events to one API call as stated further below.

For advanced integrations, where a single shell script can contain multiple commands, you can install a {SmartProxy} shellhooks plugin that exposes executables using a REST HTTP API.

You can then configure a webhook to reach out to a {SmartProxy} API to run a predefined shellhook.
The shellhook can for example contain commands or edit files.

> You must place the scripts in `/var/lib/foreman-proxy/shellhooks` as executables with only alphanumeric characters and underscores in the name.

Reviewer 2: ```suggestion
You must place your executable scripts in `/var/lib/foreman-proxy/shellhooks` with only alphanumeric characters and underscores in their name.
```

You can pass input to shellhook script through the webhook payload.
This input is redirected to standard input of the shellhook script.

> You can pass arguments to shellhook script using HTTP headers in format X-Shellhook-Arg-1 to X-Shellhook-Arg-99.

Reviewer 2: IMHO needs some form of highlighting fot your example format.

Author: Good catch! Adding monospace.

For more information on passing arguments to shellhook script, see:

* xref:passing-arguments-to-shellhook-script-using-webhooks_{context}[]
* xref:passing-arguments-to-shellhook-script-using-curl_{context}[]

The HTTP method must be POST.
An example URL would be: `\https://{smartproxy-example-com}:{smartproxy_port}/shellhook/My_Script`.

[NOTE]
====
Unlike the `shellhooks` directory, the URL must contain `/shellhook/` in singular to be valid.
====

You must enable *{SmartProxy} Authorization* for each webhook connected to a shellhook to enable it to authorize a call.

> Standard output and standard error output are redirected to the {SmartProxy} logs as messages with debug or warning levels respectively.

Reviewer 2: logs files of your {SmartProxyServer}?

Author: I think {SmartProxy} logs is fine in this case as it is shorter.

Scripts must be placed in `/var/lib/foreman-proxy/shellhooks` as executables with only alphanumeric characters and underscores in the name.

Because the shellhook script only uses command line arguments, the HTTPS payload can be empty (the *Empty Payload* webhook template).

> You can provide command line arguments using X-Shellhook-Arg-1 _to_ N.

Reviewer 3: Should this mention `N = 99`?

Author: Added :+1:
